torch/fx/graph_module.py

```diff
@@ -276,6 +264,113 @@ def _gen_model_repr(module_name: str, module: torch.nn.Module) -> Optional[str]:
+
+        Args:
+            target: The fully-qualified string name of the new submodule
+            m: The submodule itself; the actual object we want to
+                install in the current GraphModule
+        """
+        *prefix, field = target.split('.')
+
+        for item in prefix:
+
+            t = getattr(self, item, None)
+
+            if t is None:
+                t = torch.nn.Module()
+                setattr(self, item, t)
+
+            self = t
+
+        setattr(self, field, m)
+
+    def delete_submodule(self, target: str) -> bool:
+        """
+        Deletes the given submodule from ``self``.
+
+        The module will not be deleted if ``target`` is not a valid
+        target. This method will also delete parent submodules if they
+        if become unused after deleting their child. For example,
+        ``target=foo.baz.bar`` will first cause ``bar`` to be deleted.
+        If deleting ``bar`` means that ``foo.baz`` is unused, then
+        ``foo.baz`` will be deleted. If deleting ``baz`` means that
+        ``foo`` is unused, ``foo`` will be deleted as well. The most
+        deeply-nested submodule will always be deleted, regardless of
+        whether or not it contains other submodules. (So, in the above
+        example, ``foo.bar.baz`` will be deleted even if it contains
+        ``foo.bar.baz.qux``.)
+
+        Args:
+            target: The fully-qualified string name of the new submodule
+
+        Returns:
+            bool: Whether or not the target string referenced a
+                submodule we want to delete. A return value of ``False``
+                could mean either a) the submodule doesn't exist or
+                b) the submodule is an intermediary submodule that
+                contains other submodules and thus can't be safely
+                deleted.
+        """
+
+        def delete_nested(mod: torch.nn.Module, target: str) -> bool:
+            item, *suffix_list = target.split('.', 1)
+
+            if not hasattr(mod, item):
+                return False
+
+            # If we've reached the final item in `target`, e.g. `baz` of
+            # `foo.bar.baz`
+            if not suffix_list:
+                delattr(mod, item)
+                return True
+
+            submod = getattr(mod, item)
+
+            if not delete_nested(submod, "".join(suffix_list)):
+                return False
+
+            # We can delete the current module if it doesn't have any
+            # other child modules. `named_children` returns an iterator,
+            # which is why we want to check using a try-except block
+            has_other_submodules = True
+            try:
+                next(submod.named_children())
+            except StopIteration:
+                has_other_submodules = False
+
+            if has_other_submodules:
+                return False
+
+            delattr(mod, submod)
+            return True
+
+        return delete_nested(self, target)
```

Comment:
I don't think the logic here is quite correct: a `Module` could be used if any one of the following is True:
1) It has children that are used
2) Its forward is called directly via a `call_module` node
3) It has a non-Module attribute that is used from a `get_attr` node

I think this function should instead just delete the `Module` that was specified by `target`. Then optionally there can be a separate function that accounts for the 3 use scenarios above and GCs modules. (BTW I think that last part can be done by calling `GraphModule.named_modules()`, iterating over the `Nodes`, and accounting for what's used, rather than a recursive algorithm)